Fast tmx (added culling for Tile Map)

dabingnn opened the pull request:
Please do not merge it now. 
Feedbacks and comments are appreciated.
Performance tests and improvements is needed.

Commit:
issue #4437: add new tmx.

Commit:
issue #4437: add test and fix transform.

Commit:
issue #4437: add dirty tag.

Commit:
issue #4437: fix some bugs.

Commit:
issue #4437: fix crash when doesn't support VAO, and make calculation more precisely.

Commit:
Merge branch 'v3' of https://github.com/cocos2d/cocos2d-x into fast_tmx

Commit:
issue #4437: config for android and linux.

Commit:
issue #4437: add vertex z support, fix some bugs.

Commit:
issue #4437: change name TMXLayer2 and TMXTiledMap2 to FastTMXLayer and FastTMXTiledMap, change Point to Vec2

dabingnn committed:
refactor fast tmx rendering

dabingnn committed:
Merge branch 'fast_tmx_copy' into fast_tmx

Conflicts:
	cocos/2d/CCFastTMXLayer.cpp

dabingnn committed:
use quad command for tile map rendering

dabingnn committed:
new version of rendering for tile map

dabingnn committed:
clean up Fast TMX code

dabingnn committed:
clean up Fast TMX code2

dabingnn committed:
Merge branch 'v3' into fast_tmx

Conflicts:
	build/cocos2d_libs.xcodeproj/project.pbxproj

dabingnn committed:
add fast tmx files to Xcode project file

dabingnn committed:
using stable sort for quads

gelldur commented on cocos/2d/CCFastTMXLayer.cpp:
```diff
@@ -0,0 +1,777 @@
+    {
+        if(y<0 || y >= _layerSize.height)
+            continue;
+        for (int x = visibleTiles.origin.x - tilesOverX; x < visibleTiles.origin.x + visibleTiles.size.width + tilesOverX; ++x)
```
Why this for is so big :/ could you refactor it?

gelldur commented on cocos/2d/CCFastTMXLayer.cpp:
```diff
@@ -0,0 +1,777 @@
+            {
+                break;
+            }
+        } // for x
```
If you refactor for you don;t have to do such things

gelldur commented on cocos/2d/CCFastTMXLayer.cpp:
```diff
@@ -0,0 +1,777 @@
+
+    Size screenSize = Director::getInstance()->getWinSize();
+
+    switch (_layerOrientation)
```
default missing. Assert or something?

minggo replied:
It is fixed.

gelldur commented on cocos/2d/CCFastTMXLayer.cpp:
```diff
@@ -0,0 +1,777 @@
+void FastTMXLayer::parseInternalProperties()
+{
+    auto vertexz = getProperty("cc_vertexz");
+    if (!vertexz.isNull())
```
why you don't do such thing

if(vertexz.isNull() == nullptr)
{
    return;
}

you will save indention

gelldur commented on cocos/2d/CCFastTMXLayer.cpp:
```diff
@@ -0,0 +1,777 @@
+    TMXTileFlags currentFlags;
+    int currentGID = getTileGIDAt(tileCoordinate, &currentFlags);
+    
+    if (currentGID != gid || currentFlags != flags)
```
again you can save indention

gelldur commented on cocos/2d/CCFastTMXTiledMap.cpp:
```diff
@@ -0,0 +1,259 @@
+{
+    Size size = layerInfo->_layerSize;
+    auto& tilesets = mapInfo->getTilesets();
+    if (tilesets.size()>0)
```
you can save a lot of indentions

dabingnn committed:
add comments

dabingnn committed:
adjust some code to reduce indentions

dabingnn committed:
use custom command for rendering

dabingnn committed:
use custom command for fast_tmx rendering

dabingnn committed:
improvement for fast tmx rendering

dabingnn committed:
roll back quadcommand

dabingnn committed:
roll back fps compute logic
fix crash bug

dabingnn committed:
clean up fast tmx code

dabingnn committed:
fix wrong indentions

minggo commented on cocos/2d/CCFastTMXLayer.h:
```diff
@@ -0,0 +1,264 @@
+    int getTileGIDAt(const Vec2& tileCoordinate, TMXTileFlags* flags = nullptr);
+    CC_DEPRECATED_ATTRIBUTE int tileGIDAt(const Vec2& tileCoordinate, TMXTileFlags* flags = nullptr){
+        return getTileGIDAt(tileCoordinate, flags);
+    };
```
Why new class need deprecated functions?

minggo commented on cocos/2d/CCFastTMXTiledMap.cpp:
```diff
@@ -0,0 +1,259 @@
+    if (tilesets.size()>0)
+    {
+        TMXTilesetInfo* tileset = nullptr;
+        for (auto iter = tilesets.crbegin(); iter != tilesets.crend(); ++iter)
```
Could use for loop.

minggo commented on cocos/2d/CCFastTMXTiledMap.h:
```diff
@@ -0,0 +1,221 @@
+ */
+
+/** Possible orientations of the TMX map */
+enum
```
Please use `enum class`.

minggo commented on cocos/2d/CCFastTMXTiledMap.h:
```diff
@@ -0,0 +1,221 @@
+    FastTMXOrientationHex,
+
+    /** Isometric orientation */
+    FastTMXOrientationIso,
```
Elements should use capital words, and use `_` to connect words.

minggo commented on cocos/2d/CCFastTMXTiledMap.h:
```diff
@@ -0,0 +1,221 @@
+THE SOFTWARE.
+****************************************************************************/
+#ifndef __CCTMX_TILE_MAP2_H__
+#define __CCTMX_TILE_MAP2_H__
```
Don't match file name.

dabingnn committed:
windows works for fast tmx

minggo commented on cocos/2d/CCFastTMXTiledMap.h:
```diff
@@ -0,0 +1,221 @@
+    };
+    
+    /** properties */
+    inline ValueMap& getProperties() { return _properties; };
```
const?

minggo commented on cocos/2d/CCFastTMXTiledMap.h:
```diff
@@ -0,0 +1,221 @@
+objectGroup->getProperty(name_of_the_property);
+object->getProperty(name_of_the_property);
+
+@since v0.8.1
```
since 3.2

dabingnn committed:
remove deprecated function
rename enum to static const member of class
remove hex tile map(we do not support it yet)

dabingnn committed:
refactor FastTMXTiledMap::tilesetForLayer to a more elegant implementation

dabingnn committed:
refactor updateTiles

minggo commented:
Please fix conflict.
Thanks.

dabingnn committed:
Merge branch 'v3' into fast_tmx

Conflicts:
	cocos/2d/cocos2d.vcxproj.filters

dabingnn committed:
fix merge conflitcts on windows
